### PR TITLE
Implement Azure Timeseries Insights Connector

### LINKED ISSUE #21315: Implement Ballerina Azure Insights connector
**Description:**
Continuation of POC done in https://github.com/ballerina-platform/ballerina-lang/issues/21183 and add test cases for implemtation

### PR DESCRIPTION
**Pull request by niveathika**
## Purpose
Azure TSI exposed API's to query environment, however, the complex JSON is accepted and returned in request and response and it is not very user-friendly

## Goals
Ballerina Azure time-series connectors attempt to improve usability when connecting to TSI REST API through ballerina by the use of mapping JSON to ballerina record and validations.
Fix https://github.com/ballerina-platform/ballerina-lang/issues/21315

## Approach
1. Mapped JSON to relevant records
2. Introduced 2 clients to connect to REST API group

## User stories
https://groups.google.com/forum/?hl=en#!topic/ballerina-dev/Y7Fjjy8aDZM

## Release note
Ballerina Azure Timeseries Insights connector

## Documentation
Attached in the PR

## Training
N/A

## Certification
N/A

## Marketing
N/A

## Automation tests
 - Unit tests - N/A
 - Integration tests - Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Samples
```ballerina
import ballerina/log;
import ballerinax/azure.timeseries;
azure.timeseries:ConnectionConfiguration connConfig = {
    tenantId: <tenant Id>,
    clientId: <client id>,
    clientSecrect: <client secrect>
};

public function main() {
    
    azure.timeseries:EnvironmentClient environmentClient = new(<envFQDN>, connConfig);
    
    EventsRequest eventsRequest = {
        searchSpan: {
            'from: {
                dateTime: "2019-12-30T00:00:00.000Z"
            },
            to: {
                dateTime: "2021-12-30T00:00:00.000Z"
            }
        },
        predicate: {
            eq: {
        'left: {
            property: "lattitude",
            'type: "Double"
        },
        'right: 3.14
            }
        },
        top: {
            sort: [
                {
                    input: {
                        builtInProperty: "$ts"
                    },
                    'order: ORDER_DESCENDING
                }
            ],
            count: 5
        }
    };

    var response = environmentClient->getEvents(eventsRequest);
    if (response is EventsResponse) {
        io:println("Events " + response.events);
    } else {
        io:println(response.detail()?.message.toString());
    }
}

```
## Related PRs
N/A

## Migrations (if applicable)
N/A

## Test environment
jBallerina-1.2.0-rc1
 
## Learning
N/A

### CODE REVIEW

File: src/azure.timeseries/constants.bal [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+const string ERROR_CODE = "(ballerinax/azure.timeseries)ResponseError";
+
+const string VERSION = "?api-version=2016-12-12";
+
+const string INSIGHTS_BASE_URL = "https://api.timeseries.azure.com/environments";
+const string AZURE_LOGIN_BASE_URL = "https://login.microsoftonline.com/";
+const string AZURE_TSI_DEFAULT_SCOPE = "https://api.timeseries.azure.com//.default";
```
Comment (kalaiyarasiganeshalingam) [inline src/azure.timeseries/constants.bal]: Can you add new line?

File: src/azure.timeseries/endpoint.bal [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            auth: {
+                authHandler: bearerHandler
+            },
+            http1Settings: {
+                proxy: connConfig.proxyConfig
+            }
+        });
+    }
+
+
```
Comment (kalaiyarasiganeshalingam) [inline src/azure.timeseries/endpoint.bal]: Can you remove the extra line?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                  CLIENT_ID: ${{ secrets.CLIENT_ID }}
+                  CLIENT_SECRET: ${{ secrets.CLIENT_SECRET }}
+                  
+            - name: Ballerina Push
+              uses: ballerina-platform/ballerina-action/@master
+              with:
+                  args:
+                      push -a
+              env:
+                  BALLERINA_CENTRAL_ACCESS_TOKEN: ${{ secrets.BALLERINA_CENTRAL_ACCESS_TOKEN }}
```
Comment (wggihan) [inline .github/workflows/release.yml]: Need a new line here

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+script:
+  - wget https://product-dist.ballerina.io/downloads/1.2.0/ballerina-linux-installer-x64-1.2.0.deb
+  - sudo dpkg -i ballerina-linux-installer-x64-1.2.0.deb
+  - sudo apt-get install -f
+  - ballerina --version
+  - ballerina build -c --skip-tests azure.timeseries
```
Comment (wggihan) [inline .travis.yml]: Add a new line here.

File: src/azure.timeseries/constants.bal (hunk 1 above)
Comment (niveathika) [reply src/azure.timeseries/constants.bal]: Fixed in 8f58a0f

File: src/azure.timeseries/endpoint.bal (hunk 2 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in 8f58a0f

File: .github/workflows/release.yml (hunk 3 above)
Comment (niveathika) [reply .github/workflows/release.yml]: Fixed in 8f58a0f

File: .travis.yml (hunk 4 above)
Comment (niveathika) [reply .travis.yml]: Fixed in 8f58a0f

File: README.md [hunk 5]
```diff
@@ -1,2 +1,99 @@
-# module-azuretimeseries
-Ballerina connector for Azure Time-series Inisights
+## Module Overview
+
+The Azure Insights connector allows you to query Azure Time-series Insights through the  REST API. It also allows you access metadata on environments along with querying for events and aggregates. It handles OAuth 2.0 authentication.
```
Comment (praneesha) [inline README.md]:
```suggestion
The Azure Insights connector allows you to query Azure Time-series Insights through the REST API. It also allows you to access metadata on environments along with querying for events and aggregates. It handles OAuth 2.0 authentication.
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,99 @@
...
+
+## Azure Insights Operations
+
+The `ballerinax/azure.timeseries` module allows you to perform following operations,
```
Comment (praneesha) [inline README.md]:
```suggestion
The `ballerinax/azure.timeseries` module allows you to perform the following operations.
```

File: README.md [hunk 7]
```diff
@@ -1,2 +1,99 @@
...
+
+- Get all available environment meta data
+- Get Environment availability
```
Comment (praneesha) [inline README.md]:
```suggestion
- Get environment availability
```

File: README.md [hunk 8]
```diff
@@ -1,2 +1,99 @@
...
+- Get Environment metadata
```
Comment (praneesha) [inline README.md]:
```suggestion
- Get environment metadata
```

File: README.md [hunk 9]
```diff
@@ -1,2 +1,99 @@
...
+## Feature Overview
+
+### Azure Insights Clients
+
+There are 2 clients provided by Ballerina to interact with different API groups of the Timeseries Insights REST API. 
+
+1. **azure.timeseries:InsightsClient** - This client is the top-most client in the Timeseries module. 
+This can be used to get the relevant metadata associated with the environments available to the user.
+
+2. **azure.timeseries:EnvironmentClient** - Client that can be used to get metadata/query a specific environment.
```
Comment (praneesha) [inline README.md]:
```suggestion
2. **azure.timeseries:EnvironmentClient** - This client can be used to get metadata or query a specific environment.
```

File: README.md [hunk 10]
```diff
@@ -1,2 +1,99 @@
...
+
+## Getting Started
+
+### Prerequisites
+Download and install [Ballerina](https://ballerinalang.org/downloads/).
+
+### Pull the Module
+Execute the below command to pull the Azure.Timeseries module from Ballerina Central:
```
Comment (praneesha) [inline README.md]:
```suggestion
Execute the below command to pull the `Azure.Timeseries` module from Ballerina Central:
```

File: src/azure.timeseries/Module.md [hunk 11]
```diff
@@ -0,0 +1,143 @@
+Connects to Azure Time-series Insights from Ballerina.
+
+# Module Overview
+
+The Azure Insights connector allows you to query Azure Time-series Insights through the  REST API. It also allows you access metadata on environments along with querying for events and aggregates. It handles OAuth 2.0 authentication.
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
The Azure Insights connector allows you to query Azure Time-series Insights through the  REST API. It also allows you to access metadata on environments along with querying for events and aggregates. It handles OAuth 2.0 authentication.
```

File: src/azure.timeseries/Module.md [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+
+## Azure Insights Operations
+
+The `ballerinax/azure.timeseries` module allows you to perform following operations,
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
The `ballerinax/azure.timeseries` module allows you to perform the following operations.
```

File: src/azure.timeseries/Module.md [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+- Get all available environment meta data
+- Get Environment availability
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
- Get environment availability
```

File: src/azure.timeseries/Module.md [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+- Get Environment metadata
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
- Get environment metadata
```

File: src/azure.timeseries/Module.md [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+| Azure Time-series Insights API          | 2016-12-12                  |
+
+## Azure Insights Clients
+
+There are 2 clients provided by Ballerina to interact with different API groups of the Timeseries Insights REST API. 
+
+1. **azure.timeseries:InsightsClient** - This client is the top-most client in the Timeseries module. 
+This can be used to get the relevant metadata associated with the environments available to the user.
+
+2. **azure.timeseries:EnvironmentClient** - Client that can be used to get metadata/query a specific environment.
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
2. **azure.timeseries:EnvironmentClient** - This client can be used to get metadata or query a specific environment.
```

File: src/azure.timeseries/Module.md [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+## Sample
+
+### Create Environment client
```
Comment (praneesha) [inline src/azure.timeseries/Module.md]:
```suggestion
### Create an Environment client
```

File: src/azure.timeseries/endpoint.bal [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            auth: {
+                authHandler: bearerHandler
+            },
+            http1Settings: {
+                proxy: connConfig.proxyConfig
+            }
+        });
+    }
+
+    # Get details of all the available environments
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Get details of all the available environments.
```

File: src/azure.timeseries/endpoint.bal [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+                }
+                return processResBodyError(jsonResponse);
+            }
+            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response");
+        } else {
+            return processResError(httpResponse);
+        }
+    }
+
+    # Initiate and get an environment Client
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Initiate and get an environment Client.
```

File: src/azure.timeseries/endpoint.bal [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+            auth: {
+                authHandler: bearerHandler
+            },
+            http1Settings: {
+                proxy: connConfig.proxyConfig
+            }
+        });
+    }
+
+    # Get the availability details of the environment
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Get the availability details of the environment.
```

File: src/azure.timeseries/endpoint.bal [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    #
+    # + return - AvailabiltyResponse record if successful else an error
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # + return - the `AvailabiltyResponse` record if successful or an error
```

File: src/azure.timeseries/endpoint.bal [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+                }
+                return processResBodyError(jsonResponse);
+            }
+            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response");
+        } else {
+            return processResError(httpResponse);
+        }
+    }
+
+    # Get the Environment Metadata i.e. all properties metadata for a specific timeperiod
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Get the Environment Metadata i.e., all properties metadata for a specific time period.
```

File: src/azure.timeseries/endpoint.bal [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+                return processResBodyError(jsonResponse);
+
+            }
+            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response");
+        } else {
+            return processResError(httpResponse);
+        }
+    }
+
+    # Get the Events for a specific time interval based on any filter if needed
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Get the Events for a specific time interval based on any filter if needed.
```

File: src/azure.timeseries/endpoint.bal [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+            }
+            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response");
+        } else {
+            return processResError(httpResponse);
+        }
+    }
+
+    # Get the Events for a specific time interval based on any filter if needed 
+    #
+    # + eventRequest - EventRequest Record
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # + eventRequest - the `EventRequest` Record
```

File: src/azure.timeseries/endpoint.bal [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+    # + return - Event Response record which will contain returned events and metadata or error
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # + return - Event Response record, which will contain returned events and metadata or error
```

File: src/azure.timeseries/endpoint.bal [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+
+        if (httpResponse is http:Response) {
+            var jsonResponse = httpResponse.getJsonPayload();
+            if (jsonResponse is json) {
+                if (httpResponse.statusCode == http:STATUS_OK) {
+                    return createEventsResponse(jsonResponse);
+                }
+                return processResBodyError(jsonResponse);
+            }
+            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response");
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
            return error(ERROR_CODE, message = "Error occurred while accessing the JSON payload of the response.");
```

File: src/azure.timeseries/endpoint.bal [hunk 26]
```diff
@@ -0,0 +1,218 @@
...
+        } else {
+            return processResError(httpResponse);
+        }
+    }
+
+    # Get the aggregates of the data based on aggregation rules within a time interval
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # Get the aggregates of the data based on aggregation rules within a time interval.
```

File: src/azure.timeseries/endpoint.bal [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    #
+    # + aggregateRequest - Aggregate Request Record
+    # + return - Aggregate Response record which will contain aggregates events
```
Comment (praneesha) [inline src/azure.timeseries/endpoint.bal]:
```suggestion
    # + return - Aggregate Response record, which will contain aggregates events
```

Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 26 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 25 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 24 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 23 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: README.md (hunk 5 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: README.md (hunk 6 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: README.md (hunk 7 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: README.md (hunk 8 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: README.md (hunk 9 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: README.md (hunk 10 above)
Comment (niveathika) [reply README.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 11 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 12 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 13 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 15 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 14 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/Module.md (hunk 16 above)
Comment (niveathika) [reply src/azure.timeseries/Module.md]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 17 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 18 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 19 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 20 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 22 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: src/azure.timeseries/endpoint.bal (hunk 21 above)
Comment (niveathika) [reply src/azure.timeseries/endpoint.bal]: Fixed in cbf5566

File: README.md [hunk 28]
```diff
@@ -1,2 +1,100 @@
...
+- Get environment availability
+- Get environment metadata
+- Query environment for events
+- Query environment for aggregates
+
+## Compatibility
+
+|                             |       Versions              |
+|:---------------------------:|:---------------------------:|
+| Ballerina Language          | 1.1.x                       |
```
Comment (Maninda) [inline README.md]:
```suggestion
| Ballerina Language          | 1.2.x                       |
```

File: README.md [hunk 29]
```diff
@@ -1,2 +1,100 @@
...
+
+1. **azure.timeseries:InsightsClient** - This client is the top-most client in the Timeseries module. 
+This can be used to get the relevant metadata associated with the environments available to the user.
+
+2. **azure.timeseries:EnvironmentClient** - This client can be used to get metadata or query a specific environment.
+
+## Getting Started
+
+### Prerequisites
+Download and install [Ballerina](https://ballerinalang.org/downloads/).
```
Comment (Maninda) [inline README.md]:
```suggestion
Download and install [Ballerina](https://ballerina.io/downloads/).
```

File: README.md [hunk 30]
```diff
@@ -1,2 +1,100 @@
...
+```
+
+## Sample
+
+The EnvironmentClient `getEvents()` remote function can be used to query events available in the environment.
+
+```ballerina
+import ballerina/log;
+import ballerinax/azure.timeseries;
+azure.timeseries:ConnectionConfiguration connConfig = {
```
Comment (Maninda) [inline README.md]:
```suggestion
timeseries:ConnectionConfiguration connConfig = {
```

File: README.md [hunk 31]
```diff
@@ -1,2 +1,100 @@
...
+    tenantId: <tenant Id>,
+    clientId: <client id>,
+    clientSecrect: <client secrect>
+};
+
+public function main() {
+    
+    azure.timeseries:EnvironmentClient environmentClient = new(<envFQDN>, connConfig);
```
Comment (Maninda) [inline README.md]:
```suggestion
    timeseries:EnvironmentClient environmentClient = new(<envFQDN>, connConfig);
```

**Review by daneshk — APPROVED** [review]

File: src/azure.timeseries/Module.md [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+- Get environment availability
+- Get environment metadata
+- Query environment for events
+- Query environment for aggregates
+
+## Compatibility
+
+|                             |       Versions              |
+|:---------------------------:|:---------------------------:|
+| Ballerina Language          | 1.1.x                       |
```
Comment (wggihan) [inline src/azure.timeseries/Module.md]:
```suggestion
| Ballerina Language          | 1.2.x                       |
```

**Pull request merged**